File: Searching and Sorting/Ternary Search.py
def ternary_search(arr, first, last, key):
    while(first>=last):
        mid1 = first + ((last-first)//3)
        mid2 = last - ((last-first)//3)

        if(key == arr[mid1]):
            return mid1

        elif(key == arr[mid2]):
            return mid2

        elif(key < arr[mid1]):
            return ternary_search(arr, first, mid1-1, key)

        elif(key > arr[mid2]):
            return ternary_search(arr, mid2+1, last, key)

        elif(key < arr[mid2] and key > arr[mid1]):
            return ternary_search(arr, mid1+1, mid2-1, key)

    return -1


arr = [10, 20, 30, 40, 50]
key = 10

result = ternary_search(arr, 0, len(arr), key)

if result == -1:
    print("not found")
else:
    print("Element found at", result+1)

File: Searching/cpp/ternarySearch.cpp
#include <bits/stdc++.h>

using namespace std;

#define loop(i, a, b) for(int i = a; i < b; i++)
#define autoLoop(a, n) for(auto& a : n)
#define sze(x) (int)(x).size()

typedef long long ll;
typedef pair<int, int> pii;
typedef vector<int> vi;

void insertArray(vector<int> &nums) {
    int size, num;
    cin >> size;
    for (int i=0; i<size; i++) {
        cin >> num;
        nums.push_back(num);
    }
}

int ternarySearch(vector<int> &nums, int first, int last, int findNum) {
    if (first <= last) {
        int mid1 = first + (last-first)/3, mid2 = last - (last-first)/3;
        if (findNum == nums[mid1])
            return mid1;
        else if(findNum == nums[mid2]) 
            return mid2;
        else if (findNum < nums[mid1]) {
            return ternarySearch(nums, 0, mid1-1, findNum);
        }
        else if (findNum > nums[mid1] && findNum < nums[mid2]) {
            return ternarySearch(nums, mid1+1, mid2-1, findNum);
        }
        else if (findNum > nums[mid2]){
            return ternarySearch(nums, mid2+1, nums.size(), findNum);
        }
    }
    else
        return -1;
}

main() {
    vector <int> nums;
    int findNum;
    insertArray(nums);
    cin >> findNum;
    int first = 0, last = nums.size()-1;
    int ret = ternarySearch(nums, first, last, findNum);
    if (ret == -1) 
        cout << "not found";
    else
        cout << "element found at " << ret << " position";
}